ProcessList.c

```diff
@@ -415,64 +416,58 @@ static void ProcessList_buildTree(ProcessList* this) {
-            process->tree_left = node_counter++;
-            process->tree_index = node_index++;
-            Vector_add(this->processes2, process);
-            Hashtable_put(this->displayTreeSet, process->tree_index, process);
-            ProcessList_buildTreeBranch(this, process->pid, 0, 0, direction, process->showChildren, &node_counter, &node_index);
-            process->tree_right = node_counter++;
-            break;
-         }
-      }
+   int size = Vector_size(this->processes);
+   for (int i = 0; i < size; i++) {
+      Process* process = (Process*)Vector_get(this->processes, i);
 
-      // There should be no loop in the process tree
-      assert(i < size);
+      pid_t ppid = Process_getParentPid(process);
+      bool isRoot = false;
+
+      // If PID corresponds with PPID (e.g. "kernel_task" (PID:0, PPID:0)
+      // on Mac OS X 10.11.6) regard this process as root.
+      if (process->pid == ppid)
+         isRoot = true;
+
+      // On Linux both the init process (pid 1) and the root UMH kernel thread (pid 2)
+      // use a ppid of 0. As that PID can't exist, we can skip searching for it.
+      if (!ppid)
+         isRoot = true;
+
+      // Lookup the parent via the processTable hashtable not modified in buildTree
+      if (ProcessList_findProcess(this, ppid) == NULL)
+         isRoot = true;
+
+      // If parent not found, then construct the tree with this node as root
+      if (isRoot) {
+         process = (Process*)Vector_get(this->processes, i);
+         process->indent = 0;
+         process->tree_depth = 0;
+         process->tree_left = node_counter++;
+         process->tree_index = node_index++;
+         Vector_add(this->displayList, process);
+         Hashtable_put(this->displayTreeSet, process->tree_index, process);
+         ProcessList_buildTreeBranch(this, process->pid, 0, 0, direction, process->showChildren, &node_counter, &node_index);
+         process->tree_right = node_counter++;
+         continue;
+      }
    }
 
-   // Swap listings around
-   Vector* t = this->processes;
-   this->processes = this->processes2;
-   this->processes2 = t;
-
    // Check consistency of the built structures
-   assert(Vector_size(this->processes) == vsize); (void)vsize;
-   assert(Vector_size(this->processes2) == 0);
+   assert(Vector_size(this->displayList) == vsize); (void)vsize;
 }
 
-void ProcessList_sort(ProcessList* this) {
+void ProcessList_updateDisplayList(ProcessList* this) {
    if (this->settings->ss->treeView) {
       ProcessList_updateTreeSet(this);
-      Vector_quickSortCustomCompare(this->processes, ProcessList_treeProcessCompare);
+      Vector_quickSortCustomCompare(this->displayList, ProcessList_treeProcessCompare);
    } else {
-      Vector_insertionSort(this->processes);
+      if (this->needsSort)
+         Vector_insertionSort(this->processes);
+      Vector_prune(this->displayList);
+      int size = Vector_size(this->processes);
+      for (int i = 0; i < size; i++) {
```

Comment:
`size_t`, also could drop those brace

Reply:
Does it really make sense to use `size_t` for a loop from 0 to `Vector_size(this->processes)` given that `Vector_size` is defined as returning `int`?